Fixed point celsius type

X-Ryl669 opened the pull request:
### Description

This changes the `celsius_t` type from int16_t to a fixed point number. This allows to gain back the precision lost in the recent `celsius_t` type refactor, while still keeping decent performance (most operations are native integer operation on fixed point format) instead of emulated floating point.

There's probably another pass to do to change the last floating point computation in the code to fixed point (for temperature), but the required precision is not clear in the current algorithm, so I haven't changed the code where I wasn't sure.

### Requirements

#21494 for unit tests (the fixed point template class has numerous unit tests to ensure it's not doing wrong computations) 

### Benefits

Instead of integer only temperature in Marlin, you're getting back the 0.15°C precision here (the fixed point format is 1.9.6), so there is 64 units of division of a degree. This should appear in both serial output and while computing them. 
The drawback is a limitation of the temperature range to [+511°C    -512°C] but I think it's acceptable.

### Configurations

N/A

### Related Issues

#21374 Last comment

thinkyhead committed:
Suppress redundant serial errors

X-Ryl669 committed:
Add unit tests to Marlin

X-Ryl669 committed:
Merge branch 'bugfix-2.0.x' into addTestCode

X-Ryl669 committed:
Add fixed point class

X-Ryl669 committed:
Add fixed point for celsius_t type (should be faster)

X-Ryl669 committed:
Merge branch 'bugfix-2.0.x' into BetterTemperatureType

thinkyhead committed:
Remove extraneous 'extern' hints (#21516)

thinkyhead committed:
Fix CardReader string args

X-Ryl669 committed:
Fix build for linux. Some refactoring is still required for the numerous identic functions that are laying around here

rhapsodyv commented:
what is wrong with just plain `float`?!

ManuelMcLure commented:
> what is wrong with just plain `float`?!

Performance on CPUs without hardware floating point.

thisiskeithb commented:
> what is wrong with just plain `float`?!

Either way, as long as we can have some kind of precision with a decimal place.

It'll help with bug reports on Discord & Facebook since the stair stepping/spikes/dips in temperature graphs can look like an issue:

<p align="center"><img src="https://user-images.githubusercontent.com/13375512/113451358-ab2b3a80-93b6-11eb-8f82-06fc80f22c82.png" width="50%"></p>

thinkyhead commented:
As a standard part of such offerings, please include a report on the impact that the PR has on SRAM usage and build size, especially as it impacts the default build size.

thinkyhead commented on Marlin/src/module/planner.cpp:
```diff
@@ -1447,8 +1447,9 @@ void Planner::check_axes_activity() {
-    LIMIT(t, autotemp_min, autotemp_max);
+    float atMin = (float)autotemp_min, atMax = (float)autotemp_max;
+    float t = atMin + high * autotemp_factor; // Do we need float precision here (slower) ?
+    LIMIT(t, atMin, atMax);
```
The autotemp factor is stored as a float, so everything will be upcast to `float` in the calculation. To get rid of `float` here you will need to replace `autotemp_factor` —currently a small decimal number— with a scaled integer, and then down-scale the result of the multiplication.

X-Ryl669 replied:
I can convert everything to celsius_t here and it'll be computed with 6 bits of precision for the small value (faster than float, since it's using HW integer multiplier). But the question is the range here. If `autotemp_factor` is very small we'll lose precision. How small is it ? Smaller than 0.15  (in that case, using float will be better) or much larger ?

thinkyhead replied:
Users can set `autotemp_factor` to any legal `float` value, and its value should be preserved with a resolution of 0.001 to avoid annoying them.

thinkyhead commented on Marlin/src/module/temperature.cpp:
```diff
@@ -583,7 +591,7 @@ volatile bool Temperature::raw_temps_ready = false;
 
         // Get the current temperature and constrain it
-        current_temp = GHV(temp_chamber.celsius, temp_bed.celsius, temp_hotend[heater_id].celsius);
+        current_temp = (float)GHV(temp_chamber.celsius, temp_bed.celsius, temp_hotend[heater_id].celsius);
```
Compiler warning?

X-Ryl669 replied:
The right part is a celsius_t  but the current_temp is a float (I've kept it as float because I did not spend time studying the required precision for the later computation). When I'll have finished the change, either it'll be celsius_t (if a huge precision is not required) or it'll stay as float. 

Don't spend too much time on those, I'm currently merging all 4 similar functions into one here so it'll save flash space.

thinkyhead committed:
initial cleanup

thinkyhead committed:
update G26 types

thinkyhead committed:
mess with fp inline specifiers

thinkyhead committed:
improve macros for strict types

thinkyhead committed:
Help compiler with new classes

thinkyhead committed:
more help

Vertabreak committed:
GT2560 V3 followup (#21512)

XDA-Bam committed:
Update u8g / AVR display timing defaults (#21518)

thinkyhead committed:
[cron] Bump distribution date (2021-04-03)

thinkyhead committed:
ibid.

X-Ryl669 commented on buildroot/share/unit-tests/TestVector.hpp:
```diff
@@ -74,58 +74,58 @@
+        if (testResult >= 0) {
+          ptrace (PT_DETACH, 0, (caddr_t) 1, 0);
+          testResult = 1;
+        }
```
I'll try to report those change in the other PR then (or I can close the other PR either way) #21494

X-Ryl669 commented on Marlin/src/core/temp_types.h:
```diff
@@ -70,5 +70,23 @@ struct celsius_t : public fixp<true, 10, 6> {
 
-  inline celsius_t& operator = (const celsius_t & other) { data = other.data; return *this; }
+  celsius_t& operator=(const celsius_t &other) { data = other.data; return *this; }
+  celsius_t& operator=(const int &val)         { data = val;        return *this; }
```
This is wrong. It's a fixed point, you must shift the value here. I don't think it's the right fix for compiler's error report.

thinkyhead replied:
I expect that I made some errors. Please follow up with any needed repairs. In this case, I assumed that your FP type automatically does the shift for an ordinary `int`. But it sounds like it is preferable to do a direct assign from an `int`…?

X-Ryl669 replied:
I don't think you need those operators. That's the job of the parent `fixp` class and it's doing it well (see the attached tests). Replicating the operators here leads to confusion and might changes the expected internal value.
You can direct assign from an int or a float and the right `fixp::operator =` will be called in consequence (so it's transparent from the calling site).

I'll fix this on Tuesday, don't worry.

thinkyhead replied:
I added these operators to get past some compile errors. I never did get past all the errors though….

X-Ryl669 commented on Marlin/src/module/temperature.cpp:
```diff
@@ -981,9 +981,9 @@ void Temperature::min_temp_error(const heater_id_t heater_id) {
-        const float hotend_ee = temp_hotend[ee].celsius,
-                    hotend_target = temp_hotend[ee].target,
-                    pid_error = hotend_target - hotend_ee;
+        const celsius_t hotend_ee = temp_hotend[ee].celsius,
```
Those are here to compute the fixed point to floating point only once. If you change them to celsius_t and use cast below, it'll cost the conversion multiple time.

thinkyhead replied:
I'm surprised the compiler in `-O2` would do something different … unless that is one of the useful side-effects of wrapping the temperature value in a struct/class.

I am still not fond of over-complicating `celsius_t`. The convenience of the nice simple type we currently have might just outweigh the added inconveniences of an over-designed complex type.

All of Marlin's internal temperature work is done with celsius units, and there is no confusion about that. Any conversion from some other unit is done immediately on input, and afterward it will only be converted to another unit for display. So there is no need to create a series of temperature types. A single simple type, perhaps able to convert to and from other units as part of the type definition, is all that is actually needed. Making the type overly-strict is not a Requirement.

As a coder who works with these types every day I would much prefer that type conversions were done as transparently as they were before, and that we did not require any extra gesticulations to get a celsius temperature in the form of a primitive type. We simply don't need all these extras attached to the type.

One of the elements of Marlin's "coding style" has been to keep it informal and non-pedantic. Keeping the `celsius_t` a simple alias for `int16_t` provides maximum convenience, and makes it much easier to manage and use the value. That factor alone outweighs formalism — or at least incomplete formalism.

When I introduced the `types.h` file with all those lengthy position types, it demonstrated just how much work and detail is required to make custom types behave in "the way that you would want and expect" at any given use. If you're expecting the fixed-to-float conversions here to be un-optimizable by the compiler then maybe the type needs more work. It really should behave transparently and not introduce more inconveniences.

X-Ryl669 replied:
There's no difference between `celsius_t` being a `int16_t` or a `fixp`. If you convert it to float in the code at multiple places, the compiler might do it once (and store the value in a register) or do it multiple time if it does not find a free register or estimate it's better this way. When I do it myself, I'm preventing the compiler to decide on its own. 
It's probable -O2 or -Os will give the same code but it can change any time depending on the code content, I haven't checked.

Converting from int to float on a CPU without a FPU is costly (it requires a recursive division). Doing it with the fixed point type requires a multiplication and a shift (better but still expensive).  

The only difference with a POD and a struct is that the conversion is implicit in the former case and I've made it explicit in the latter case. I can make it implicit, but I think it's better if the developer choose to convert so he's aware of it.

For now, I'll continue with explicit until it's compiling fine on all platform (so I can fix the code with clear intentions) and then switch back to implicit conversion for convenience.

thinkyhead replied:
Costly calculations only weigh against code size when there is a bottleneck. We don't do enough temperature conversion between types to consider it a meaningful bottleneck, so it is okay if it is a little inefficient.

thinkyhead replied:
There is of course the potential to convert all floating-point in Marlin into fixed-point, knowing the predictable ranges of certain parameters. That would be the most thoroughgoing approach, and to eschew `float` altogether when working with temperatures.

The earliest version of Prusa Firmware converted the planner to use fixed-point because it is known to have a certain limited range of bed size and micro-steps. No one has yet made that conversion in Marlin because it's a bunch of meticulous and specialized work, and it would need to be able to pick the most appropriate type for machines that have larger coordinate ranges or higher numbers of micro-steps.

X-Ryl669 replied:
I've chosen 10.6 because it allowed +511 -512 range (and 0.15°C of resolution). We could add an offset here if we decided to increase the resolution (like a FP9.7 + 192°C, to get: +448/-64°C and 0.007° of resolution). If we need more, we can choose a 32 bits type here, with: 12.20 here (it's a #ifdef away to change), so we'd get +4095/-4096 and 1µ°C of resolution, but the cost will be higher for multiplications.

thinkyhead committed:
[cron] Bump distribution date (2021-04-04)

X-Ryl669 commented:
> As a standard part of such offerings, please include a report on the impact that the PR has on SRAM usage and build size, especially as it impacts the default build size.

I will.

X-Ryl669 commented:
> what is wrong with just plain `float`?!

On platform having a FPU, it'll be better to use a float (as it was before). It's a typedef + ifdef away to switch the type here on those platforms. On the other platforms it'll be better to use fixed point.

GMagician committed:
Update Italian language (#21537)

InsanityAutomation committed:
Fix Hotend-abort-on-idle Check (#21535)

InsanityAutomation committed:
Revert "Optimized string-to-number functions" (#21532)

This reverts #21484

thinkyhead committed:
misc. pointer formatting

vyacheslav-shubin committed:
Fix FAT delete of items with long name (#21528)

marciot committed:
Touch UI Bed Mesh Screen refactor, enhancements (#21521)

- Split mesh view and edit screen into two screens
- The editor now live-updates the graphics
- Added Touch UI mesh progress feedback to `G26`
- Show positive / negative mesh values in different colors

thinkyhead committed:
Update fp_type.h

thinkyhead commented on Marlin/src/lcd/extui/ui_api.cpp:
```diff
@@ -982,21 +982,23 @@ namespace ExtUI {
     }
   }
 
   void setTargetTemp_celsius(const_float_t inval, const extruder_t extruder) {
```
By using `float` here, ExtUI is likely keeping the CocoaPress chocolate printer in mind, because it needs to be able to set temperatures in fractions of degrees.

X-Ryl669 replied:
I don't see any issue doing some computation in float where it's required. In my next changes I'll put a comment about this reason here.

thinkyhead commented on Marlin/src/module/temperature.h:
```diff
@@ -439,6 +442,7 @@ class Temperature {
 
     #if ENABLED(HAS_HOTEND)
       static temp_range_t temp_range[HOTENDS];
+      static constexpr celsius_t hotend_hysteresis = celsius_t((int)TEMP_HYSTERESIS);
```
It seems like `celsius_t(UNKNOWN)` should be fine here. If someone cheekily sets the value to `0.5` probably some `#if` statement will blow up first. But if the compiler gets to this point then I think it should be a legal value.

X-Ryl669 replied:
0.5 is a legal and accepted value here, since fixp accept both int and float here. It's just that double is not legal in template parameters (there's some template code in Marlin right now with this define).

X-Ryl669 replied:
Should be accepted in the latest commit

thinkyhead committed:
Update temperature.cpp

thinkyhead commented on Marlin/src/module/temperature.cpp:
```diff
@@ -499,8 +499,9 @@ volatile bool Temperature::raw_temps_ready = false;
-  void Temperature::PID_autotune(const_float_t target, const heater_id_t heater_id, const int8_t ncycles, const bool set_result/*=false*/) {
+  void Temperature::PID_autotune(const celsius_t desired_target, const heater_id_t heater_id, const int8_t ncycles, const bool set_result/*=false*/) {
     float current_temp = 0.0;
+    float target = (float)desired_target;
```
Does the `explicit` keyword makes this explicit cast necessary? I hope this is not needed. Call me old fashioned, but I feel that it could annoy neophyte contributors when the language appears to be obtuse, coy, or confused about the type of the lvalue, and then they will come around asking for help and use up developer time.

X-Ryl669 replied:
I understand your concern here. Yes `explicit` requires an explicit cast here. I see it as an advantage here so we can spot each time we query the conversion operator (which is not zero cost, like for casting a int to short or something). But if you feel it's too much verbose, once all the changes are applied and validated, I'll remove the explicit conversion operator so it does not appear everywhere in the code base in a single commit.

thinkyhead commented on Marlin/src/module/temperature.cpp:
```diff
@@ -1141,10 +1155,12 @@ void Temperature::min_temp_error(const heater_id_t heater_id) {
       static bool pid_reset = true;
       float pid_output = 0;
+      float temp_ch_c = (float)temp_chamber.celsius;
+      float temp_ch_target = (float)temp_chamber.target;
```
"They must be _floatified_ first. Don't you know your C++?"

thinkyhead committed:
[cron] Bump distribution date (2021-04-05)

thinkyhead committed:
write_command changes input

descipher committed:
Make DELAY_NS round up on AVR (#21546)

thinkyhead committed:
[cron] Bump distribution date (2021-04-06)

Malderin committed:
MKS LVGL UI: Display/edit fan percentage (#21544)

Co-authored-by: Scott Lahteine <github@thinkyhead.com>

rhapsodyv committed:
Fix Rumba32 variant for Marlin (#21497)

Malderin committed:
Cast celsius to int for MKS LVGL UI (#21502)

thinkyhead committed:
getHighESpeed => autotemp_task

bigtreetech committed:
BTT SKR-SE-BX (STM32H743IIT6 ARM Cortex M7) and BIQU_BX_TFT70 (#21536)

X-Ryl669 committed:
Remove useless code

ramiropolla commented:
@X-Ryl669 I don't think this is the right approach. It is a clear example of C++ OOP (_obfuscation_-oriented programming) and I would not be surprised if the compiler ended up outputting very inefficient code.

X-Ryl669 committed:
Refactor the temperature class for less code duplication

X-Ryl669 committed:
Merge branch 'bugfix-2.0.x' into BetterTemperatureType

InsanityAutomation commented:
FWIW I have a hotend here with components to go to 500c. I dont expect it to be long before industrial materials require over this as well. Even way up in that range, PEKK-C is slow printing...

X-Ryl669 committed:
Continue refactoring the temperature code so it's not as big as before and more generic

X-Ryl669 commented:
> FWIW I have a hotend here with components to go to 500c. I dont expect it to be long before industrial materials require over this as well. Even way up in that range, PEKK-C is slow printing...



> I would not be surprised if the compiler ended up outputting very inefficient code.

Honestly, there's nothing preventing from typedef'ing `celsius_t` to `float` in `temp_types.h` if you have such concerns. 

Obviously, there are limitations with fixed point compared to floating point, so I guess we can have a `#ifdef USE_FLOAT_CELSIUS` for these *specific* cases but they also have advantages (which is better than integer value for temperature, IMHO). Also, on CPU with a FPU, this should be defined anyway.

The C++ used here should be quite straightforward to change `typedef fixp<true, 10, 6> celsius_t;` to `typedef float celsius_t;` without breaking any later code (the other way is less easy, since I forbid implicit cast from celsius_t to float, but that's my internal process to spot errors in the code, I'm not sure I'll leave the cast as `explicit` in the final code).

I'm still working to fix the code to use clear units where it's mixing them. If the output of this is more compiler anger on shady code, I'm all for it. 

I've also reworked the `temperature.h/cpp` file to orthogonalize similar code (leading to 1400 removed lines at the time of writing and I haven't finished). Temperature code is full of copy & paste code (with subtle difference) that leads to duplicate binary size. I've reworked the difference in objects so I could write a single function that's behaving as before (without duplication). 

I'm not even at the step where I expect the code to compile, I'm focused on making sure I'm not breaking functionalities. 
Once this is done, I'll improve the unit tests to assert and prove I've not broken functionalities.

And then, I'll benchmark the code. I expect a huge boost from using fixed point on ARM and AVR without FPU compared to floating point. I also expect a big decrease in code size (due to the refactoring above) mainly, not from floating point vs fixp.

X-Ryl669 commented on Marlin/src/module/temperature.cpp:
```diff
@@ -768,26 +1219,70 @@ volatile bool Temperature::raw_temps_ready = false;
+    // If it's used to map an index to a fan pin (and there are multiple index leading to the same pin),
+    // then instead, we should store a bit index in the FanTemperature's above
+    /*
     #define _EFAN(B,A) _EFANOVERLAP(A,B) ? B :
```
@thinkyhead If you have some time, can you explain what this does ?
As far as I have understood, each heater can have a fan assigned to it, but different heater can have the same fan, right ?
So this is used to ensure we don't turn on a fan twice uselessly ?
How do you map from Heater to Fan index (I'm a bit lost with this macro here) ?

thinkyhead replied:
```cpp
#define _FANOVERLAP(A,B) (A##_AUTO_FAN_PIN == E##B##_AUTO_FAN_PIN)
#define _EFANOVERLAP(A,B) _FANOVERLAP(E##A,B)
::
#if HAS_AUTO_FAN_1 && !_EFANOVERLAP(1,0) . . .
#if HAS_AUTO_FAN_1 && !_FANOVERLAP(E1,0) . . .
#if HAS_AUTO_FAN_1 && !(E1_AUTO_FAN_PIN == E0_AUTO_FAN_PIN)
```

X-Ryl669 committed:
Fix a mistake on the loop limit

X-Ryl669 commented on Marlin/src/module/temperature.cpp:
```diff
@@ -3216,64 +3229,66 @@ void Temperature::tick() {
 
+    // TODO: Make an array of those, by merging them to
+    // a single state instead of so many different state
     #if HAS_TEMP_ADC_0
```
@thinkyhead : Is the order of reading the sensors important here ? 
I can probably rework these 60 lines in less than 4 here by replacing with a single state and a sub-state counter on the heater's array index. 
But this means following the heater's declaration order, that is, all hotends then temp probe, then bed, then chamber, then cooler.

ramiropolla replied:
Not the order itself, but you need some time for the adc to settle. So every time you switch adc (HAL_START_ADC) you need to wait before ACCUMULATE_ADC. Therefore one state for each.

X-Ryl669 replied:
Ok, that's the confirmation I was expecting. Thanks!

X-Ryl669 committed:
Remove useless macro now

ramiropolla commented:
> I've also reworked the `temperature.h/cpp` file to orthogonalize similar code (leading to 1400 removed lines at the time of writing and I haven't finished). Temperature code is full of copy & paste code (with subtle difference) that leads to duplicate binary size. I've reworked the difference in objects so I could write a single function that's behaving as before (without duplication).

Could you submit this refactoring as a separate pull request without any celsius_t changes? It's easier to review and test this way.

thisiskeithb commented:
Here’s a related bug report: https://github.com/MarlinFirmware/Marlin/issues/21596

X-Ryl669 committed:
Simplify the sampling loop to use heaters wherever appropriate

X-Ryl669 committed:
Fix comments

X-Ryl669 committed:
Continue working on simplifying the temperature class

X-Ryl669 commented:
@ramiropolla Once I'm done, it'll be easy to remove the celsius_t changes since it's only liked with 2 or 3 files. 
Please have a look at temp_types.h for a compile time selectable `celsius_t` type (between int16 or fixed point or float).

X-Ryl669 committed:
Rework PID autotune loop

thinkyhead commented on Marlin/src/gcode/lcd/M145.cpp:
```diff
@@ -46,8 +46,10 @@ void GcodeSuite::M145() {
-        mat.hotend_temp = constrain(parser.value_int(), EXTRUDE_MINTEMP, thermalManager.hotend_max_target(0));
+      if (parser.seenval('H')) {
+        mat.hotend_temp = parser.value_int();
+        LIMIT(mat.hotend_temp, EXTRUDE_MINTEMP, thermalManager.hotend_max_target(0));
```
Is `constrain` not usable with the new type? How can we alert future developers who try to use it that they will get back garbage?

X-Ryl669 replied:
It's probably usable. That's an error here, from the very beginning of the code. It should read:
`mat.hotend_temp = constrain(parser.value_celsius(), EXTRUDE_MINTEMP, thermalManager.hotend_max_target(0));`

thinkyhead committed:
Update parser.h

thinkyhead commented on Marlin/src/inc/Conditionals_LCD.h:
```diff
@@ -615,7 +615,8 @@
 // Helper macros for extruder and hotend arrays
-#define HOTEND_LOOP() for (int8_t e = 0; e < HOTENDS; e++)
+#define HOTEND_LOOP() for (int8_t e = 0; e < HOTENDS; e++) if (Heater & heater = Temperature::get_heater(HotEndPos0 + e))
+#define HEATER_LOOP() for (uint8_t e = 0; e < HeatersCount; e++) if (Heater & heater = Temperature::get_heater(e))
```
These changes are "too smart" and out of scope … and it looks like `&` should be `&&`. If you want a macro like `HOTEND_LOOP` that does more, don't alter `HOTEND_LOOP`. Instead define a new macro in `temperature.h` and replace `HOTEND_LOOP` in the code where needed.

X-Ryl669 replied:
Right, that's another magic macro. I needed to iterate on both the hotend index and then get a reference on a current iterated `Heater`, **without** opening a brace since the code does `HOTEND_LOOP() {...}`. The hotend index is accessible in the heater's `pos` member too so I can probably remove the `if( declare variable that always evalutate to true)` trick.

The trick is that the original `HOTEND_LOOP` is not needed anywhere else in the code now.

thinkyhead commented on Marlin/src/inc/Conditionals_post.h:
```diff
@@ -2419,6 +2419,11 @@
+struct HeaterPin {
+  pin_t pin;
+  bool  inverting;
+};
```
This should be defined elsewhere, and `typedef struct { ... } heater_pin_t;` would be more like the common idiom.

X-Ryl669 replied:
Ok. I'm seen both usage in the code (struct definition with and without typedef), and the typedef struct looks like C code to me. It has numerous drawback, like you can't have a constructor in a typedef struct since it's unamed. I'll move it back to temperature.h.

thinkyhead commented on Marlin/src/lcd/extui/ui_api.cpp:
```diff
@@ -260,7 +260,7 @@ namespace ExtUI {
     #define GET_TEMP_ADJUSTMENT(A) (float(A) / (TOUCH_UI_LCD_TEMP_SCALING))
   #else
-    #define GET_TEMP_ADJUSTMENT(A) A
+    #define GET_TEMP_ADJUSTMENT(A) float(A)
```
When most of the world (including me) is rushing towards informal scripting languages and simplified code style, why is it preferable to make this class require an explicit cast, rather than permitting the class to "do the right thing"?

Philosophically speaking, should we continue to build the code with more such 'caveats' — calling for an explicit cast for certain cases that "you'll need to know about before you enter" …?… or should we aim to make the code more non-pedantic, casual, and "script-like" to conceal the ugliness of its C++ underpinnings? I may be in a minority, but I have come to see C++ as a language designed to hide the C++ so that you can come up with your own 'super-language' on top of C++ while still producing optimal binary code in the end. Whether merited or not, changes like this feel like a move in the opposite direction.

Sure, we also see Javascript programmers running to TypeScript because they crave more formal and strict typing, and that makes sense too. What I hope for is something in the middle, where we prevent coders from doing "dumb" things but if someone forgets to `float()` the thing being assigned to a `float` the compiler will still be permissive and "do the right thing" with it.

X-Ryl669 replied:
Ok. I've heard your complaint here. I'm almost finished with my changes so I'll switch back to implicit float casting so these are not required anymore.

thinkyhead commented on Marlin/src/module/settings.cpp:
```diff
@@ -2855,11 +2814,11 @@ void MarlinSettings::reset() {
+      heater.pid.setKi(scalePID_i(PID_DEFAULT(Ki, ALIM(e, defKi))));
+      heater.pid.setKd(scalePID_d(PID_DEFAULT(Kd, ALIM(e, defKd))));
+      TERN_(PID_EXTRUSION_SCALING, heater.pid.setKc(float(PID_DEFAULT(Kc, ALIM(e, defKc)))));
+      TERN_(PID_FAN_SCALING, heater.pid.setKf(float(PID_DEFAULT(Kf, ALIM(e, defKf)))));
```
Note that without `DISTINCT_E_FACTORS` all E heaters should share the same data, and no loop should be required.

X-Ryl669 replied:
This is tough to implement. The basic idea of this PR part is to have all heater-like object be instance of a common Heater class. So even if they all share the same value, they are independent instances. What you are asking for is to use a kind of reference/pointer to the *single* value vectors here. It's possible, but it's not done now. Probably in a later PR.

thinkyhead commented on Marlin/src/module/settings.cpp:
```diff
@@ -1831,15 +1793,12 @@ void MarlinSettings::postprocess() {
       //
       {
-        PID_t pid;
+        Serialization::PIDSS pid;
```
I'm discomfited seeing double-colons creeping into the codebase. I'm not fond of them because it makes your code look like VisualBASIC, and it's me who will end up having to type them out all day.

Philosophically and stylistically speaking, will it really be better if Marlin code is modified so that everything is defined in nested namespaces with functions and variables increasingly prefixed by a hierarchy of serious-sounding namespace and class names?

X-Ryl669 replied:
Ok, no problem, I'll remove the namespace here. 


> Philosophically and stylistically speaking, will it really be better if Marlin code is modified so that everything is defined in nested namespaces with functions and variables increasingly prefixed by a hierarchy of serious-sounding namespace and class names?

I can't answer the question here. It's up to the original code writer to decide how to separate the "modules" of her code. I'm used with namespace and double colons, but I perfectly understand it's bothersome to use. So, no, there's no theoretical or better reason to use namespaces.

thinkyhead commented:
This PR is becoming too large to review and will require a lot of testing of each component. Whenever this is considered ready, please split the changes up into smaller units and submit them as separate PRs. Leave this PR open, if you wish, and it will become smaller as the separate pieces get merged.

X-Ryl669 commented:
> This PR is becoming too large to review and will require testing of each component. Whenever this is considered ready, please split the changes up into smaller units and submit them as separate PRs. Leave this PR open, if you wish, and it will become smaller as the separate pieces get merged.

Indeed. That's when you pull a spaghetto that it comes with everything attached.

X-Ryl669 committed:
Make it build by default!

X-Ryl669 committed:
Fix build for linux

X-Ryl669 committed:
Slim fast

X-Ryl669 committed:
Merge branch 'bugfix-2.0.x' into BetterTemperatureType

X-Ryl669 commented:
Bugfix:
```
RAM:   [===       ]  30.0% (used 2456 bytes from 8192 bytes)
Flash: [==        ]  21.5% (used 54720 bytes from 253952 bytes)
```
This PR:
```
RAM:   [===       ]  29.6% (used 2423 bytes from 8192 bytes)
Flash: [==        ]  15.4% (used 39028 bytes from 253952 bytes)
```
(with `celsius_t` kept as `int16`)

X-Ryl669 commented:
For what it's worth, here's a benchmark of fixed point vs floating point on AVR:
```
$ simavr --mcu atmega2560 avr.elf
Loaded 7110 .text at address 0x0
Loaded 132 .data
Hello from simavr..
Fixed point:..
  Final result: 0..
  Multiply took: 621ms..
  Final result: -400..
  Add took: 171ms..
  Final result: -400..
  Subtract took: 292ms..
  Final result: 0..
  Divide took: 6548ms..
Floating point:..
  Final result: 0..
  Multiply took: 2701ms..
  Final result: -400..
  Add took: 2631ms..
  Final result: -400..
  Subtract took: 4889ms..
  Final result: 0..
  Divide took: 4943ms..
```
The benchmarking code looks like this:
```cpp
template <typename C_t>
bool runtest() {
  // Multiplication
  {
    Test::ScopeProfile profile("Multiply");
    C_t final = C_t(0);
    for (int j = 0; j < 200; j++) {
      C_t result = C_t(1), r(Test::randomBetween(-1.0, 1.0));
      C_t g(Test::randomBetween(0, 1.0));
      C_t add(0.1);
      for (int i = 0; i < 512; i++) {
        if (i & 1) g += add; else g -= add;
        result += r * g;
      }

      if (result > 128) final++;
      else if (result < -128) final--;
    }

    fprintf(stdout, "  Final result: %d\n", (int)(float)final);
  }

  // Addition
  {
    Test::ScopeProfile profile("Add");
    C_t final = C_t(0);
    for (int j = 0; j < 400; j++) {
      C_t result = C_t(1), r(Test::randomBetween(-1.0, 1.0));
      C_t g(Test::randomBetween(0, 1.0));
      C_t add(0.1);
      for (int i = 0; i < 256; i++) {
        if (i & 1) g += add; else g -= add;
        result += r + g;
      }

      if (result > 128) final++;
      else if (result < -128) final--;
    }

    fprintf(stdout, "  Final result: %d\n", (int)(float)final);
  }

  // Subtraction
  {
    Test::ScopeProfile profile("Subtract");
    C_t final = C_t(0);
    for (int j = 0; j < 400; j++) {
      C_t result = C_t(1), r(Test::randomBetween(-1.0, 1.0));
      C_t g(Test::randomBetween(0, 1.0));
      C_t add(0.1);
      for (int i = 0; i < 512; i++) {
        if (i & 1) g += add; else g -= add;
        result += r - g;
      }

      if (result > 128) final++;
      else if (result < -128) final--;
    }

    fprintf(stdout, "  Final result: %d\n", (int)(float)final);
  }

  // Division
  {
    Test::ScopeProfile profile("Divide");
    C_t final = C_t(0);
    for (int j = 0; j < 200; j++) {
      C_t result = C_t(1), r(Test::randomBetween(0, 1.0));
      C_t g(Test::randomBetween(1.0, 32.0));
      C_t add(0.1);
      for (int i = 0; i < 512; i++) {
        if (i & 1) g += add; else g -= add;
        result += r / g;
      }

      if (result > 128) final++;
      else if (result < -128) final--;
    }

    fprintf(stdout, "  Final result: %d\n", (int)(float)final);
  }
}
```

Right now, there are 2 operations that are slower in fixed point than floating point: **division** and **converting** from/to float.
For the latter, I have some code I want to test, because the current code is doing a dumb `fixed_point_value = float * one` (with `one` being `1<<decimal position` converted to float) so it cost a floating point multiplication + a conversion to int.

I think we can avoid both here by being clever with the float format, since we are multiplying with a value that's only changing the exponent of the float here and converting to int is mainly shifting the mantissa back to our own decimal position. So instead of an expensive FPU multiplication and conversion, we can probably reduce that to a CMP + a shift + mask.

For a faster divide, honestly I've no idea here. I've implemented the Egyptian division algorithm here, but it does not perform well either. So for now, let's say we don't divide often so it's not an issue.

zeleps commented:
My two cents for the discussion:

![IMG_20210421_120239](https://user-images.githubusercontent.com/39417467/115602415-940da780-a2e7-11eb-8c0d-a7a330dac87c.jpg)

The left print is right before merging PR #21374 and the right is after the merge (print is re-sliced, but slicing parameters remained unchanged). My bed PID is performing badly, comparable to bang-band, causing noticeable banding. I tried to autotune PID, but the values I got were completely out of range. The fluctuation in temperature is noticeable in hotends as well, but it doesn't affect print quality that much.

In conclusion, it's obvious that integer temperature values are not sufficiently precise for decent PID calculations. @X-Ryl669 's suggestion of 0.15°C precision sounds adequate for my case, but imho there should be an option to adjust precision or use floating point interchangeably.

X-Ryl669 commented:
> In conclusion, it's obvious that integer temperature values are not sufficiently precise for decent PID calculations.

The PID autotune algorithm is almost *not using* the temperature values (only to compare them, it doesn't use them for any computation). So yes, it's a bit less precise but I wouldn't expect such loss in value here. That needs to be benchmarked and tested. Thanks for your contribution.

zeleps commented:
> > In conclusion, it's obvious that integer temperature values are not sufficiently precise for decent PID calculations.
> 
> The PID autotune algorithm is almost _not using_ the temperature values (only to compare them, it doesn't use them for any computation). So yes, it's a bit less precise but I wouldn't expect such loss in value here. That needs to be benchmarked and tested. Thanks for your contribution.

In temperature.cpp, get_pid_output_hotend(): 

`const float pid_error = temp_hotend[ee].target - temp_hotend[ee].celsius;`

(same goes for get_pid_output_bed())

This is greatly affected by the change to int, pid_error is 5 times more coarse now on my system, causing an oscillation of 1-2 degrees instead of a few tenths of a degree to keep the bed on target. That's why I'm getting banding :)

Calculating a float temperature value especially for the PID functions would probably solve the issue, and it's a pretty limited calculation. Maybe keeping both representations (float and int) in a struct which will replace celsius_t (celsius_t.coarse / celsius_t.fine) would be optimal: additional space required is negligible, most calculations and display functions would use the int value and where precision is affected the float would still be available. Also, calculating both values at once (float first and then truncating it to int) would save the need for code duplication. In terms of performance, it's much better than fixed point as well, float will be used in float calculations only (few and necessary), int anywhere else.

X-Ryl669 commented:
Ok, I thought you were talking about my changes. In my changes, I've replaced the pid autotune algorithm's floating point temperature by `celsius_t` since they are only compared and not used by the computation.

What you refer to is in current bugfix, not in this PR. It's used for heater's power computation during heating (cooling). This PR is instead changing the `celsius_t` type to something you can select at compilation time: either int16 (like in bugfix), or float (like in last release) or fixed point (new in this PR). 
You'll just define `FIXED_POINT_CELSIUS` or `FLOATING_POINT_CELSIUS` and everything magically improve (or worsen depending on where you put the cursor).

I'm currently working on improving the floating point conversion to fixed point (since it's something that can be improved performance wise). Then I'll try to fix all building issue for all possible configurations, and finally split the PR in different parts.

zeleps commented:
Sorry @X-Ryl669 , I had the initial celsius_t change in mind. I'll wait to see your PR completed and merged. Thanks!

rhapsodyv commented:
I took a look in the code before #21374. Before it, we have places that temp was a `int16` and places where it was a `float`. The base type was `float`.
After #21374, everything become only one type: `celsius_t`. For me, this is the main issue now. 

If we keep only one type, and change `celsius_t` to `float` we will have overhead that we hadn't before (as most of places were `int`). The same will happen with fixed point. We will add fixed point to places that we can use just `int`.

zeleps commented:
In that case @rhapsodyv, dual representation of temperature values in a struct would tackle both speed and precision issues with minimal fuss. It might be worth considering if integer operations are too frequent.

ramiropolla commented:
I've mentioned it once or twice but I can't find my references now. In my opinion, the `celsius` variable in `temp_info_t` should in fact be reverted back to `float`, and not be `celsius_t` (`int` or `fixp`). This variable contains values read from the ADC, so we should keep as much precision as possible. The calculations done in the PID autotune are not at all speed-critical, but they are precision-critical, so I think we should also keep using float for the code there.
@thinkyhead what do you think about this?

X-Ryl669 commented:
There are 2 problem here and it does not help mixing them. 

The first one is that a float is nothing. It does not say what it contains, it's implicitly cast to anything else and it's very slow to use (see benchmark above). So changing the code from float to celsius_t when the variable was actually a celsius temperature is a good move IMHO. I think Scott wanted to include his changes first so he only typedef the celsius_t to a int16 but it could have been a float either. The right move is to make it a class so that it's not implicitly converted to apple and orange without you knowing it. This PR is doing this. It makes the compiler screams when you're writing fuzzy code.

The second one is the precision of the chosen variable. int16 is a bit crude as you've spotted. My proposal is to use, at least, a fixed point on 16 bits, so we get back 0.15°C resolution instead of 1°C. Or you can still use float
There are pro and cons for theses choices which can be summed up like this:

| `celsius_t` type | Range | Precision | Code size | Speed | Works as expected |
|----------------------|----------|--------------|---------------|----------|---------------------------|
| `int16_t`            | -32768 +32768 | 1°C | smallest | very fast (50x faster than float) |   No  |
| `fp<10.6>`         | -512 +511 | 0.15°C | small | fast (at least 5x faster than float) |   TBD  |
| `fp<12.20>`       | -2048 +2047 | 1µ°C | average | quick (around 2x faster than float, expect for div) |   Likely  |
| `float`                | -1e127 +1e127 | 60n°C | larger | slow 1x |   Yes  |

In this PR, you have the possibility to select the type you want in `Configuration_adv.h`

rhapsodyv commented:
If you look at this this: https://github.com/MarlinFirmware/Marlin/pull/21374/files, _most of the temp usage were already integer_. Only a _few were float_, including calculations. I think this is the correct approach: use int when possible and a high precision when needed (float or fixed point). 

I think we need two representations, one for each usage. Handling everything as one type seems a backward step....

X-Ryl669 committed:
Add fast conversion operator to/from float - part 1

X-Ryl669 committed:
Add fast conversion operator to/from float - part 2

X-Ryl669 commented:
Now conversion is faster than "native" float to/from int because of the trick of directly changing the exponent:
```
Conversion test:..
  Final result: 12356..
  Dumb conversion took: 2754ms..  # Using (int)(float * float(1<<fractional))
  Final result: -15235..
  Conversion took: 2368ms..  # Using exponent trick for conversion
  Final result: 10637..
  Native took: 2604ms..  # Using plain (int)float conversion
  Final result: 0..
  Back conversion took: 379ms..  # Using exponent trick for (float)int conversion
  Final result: -4..
  Back conversion native took: 1217ms..  # Using plain (float)int conversion
```

X-Ryl669 commented:
> I think we need two representations, one for each usage. Handling everything as one type seems a backward step....

Your PR is mainly reverting what Scott did. I don't see any improvement here, since `typedef float celsius_float_t` is just a larger and more painful way to write float. The compiler only see a float here, so it's only bothering the developer. If you need float, use float. I've the same criticism for `const_float_t` which it more painful to use than `const float` with no advantage whatsoever (it's longer to write, force the developer to check the type, makes intellisense slower, etc...).

rhapsodyv commented:
My PR, complementing Scott PR, just make clear where we need `float` and where we need `int` for temperature values. There's no doubt now: when everyone see, he will see the places where we need high precision and the places we don't care.

We don't need make all temperature values a fixed-point, because most of the places we just need `int` (as the code ever was!!!). Changing everything  to fixed-point probably will add more overhead than remove.

So, now on, we can just change the new `celsius_float_t` to fixed-point and let the other be just `int`. This is the optimal solution.

thinkyhead commented:
So, at this point the `celsius_float_t` things are pretty well distinguished, and theoretically these could be converted to fixed-point for a variety of cases. For example, we now convert sensor readings from an ADC integer to a celsius float. Even converting this to a celsius << 3 value would improve resolution over a celsius integer by 8 times. A problem may arise later, however, in that some systems scale the stored temperature by a factor of 10 to get finer decimal adjustment, so a 32-bit fixed-point may be needed to get fine enough intervals as a replacement for that obsolete scaling.

thinkyhead committed:
Update settings.cpp

thinkyhead committed:
apply temp inline patch

thinkyhead committed:
Merge 'bugfix-2.0.x' into pr/21519

thinkyhead committed:
Merge branch 'BetterTemperatureType' of github.com:X-Ryl669/MarlinForMKS into pr/21519

thinkyhead commented:
I've done a merge to bring this up to date, and some of the changes from bugfix and choices of resolution are probably not exactly right. Please review the merge commit and patch up anything that doesn't fit.

thinkyhead committed:
fix autotemp_task
